storage/writer.go

```diff
@@ -84,30 +94,301 @@ type Writer struct {
 
 	mu  sync.Mutex
 	err error
+
+	stream storagepb.Storage_WriteObjectClient
 }
 
-func (w *Writer) open() error {
-	attrs := w.ObjectAttrs
-	// Check the developer didn't change the object Name (this is unfortunate, but
-	// we don't want to store an object under the wrong name).
-	if attrs.Name != w.o.object {
-		return fmt.Errorf("storage: Writer.Name %q does not match object name %q", attrs.Name, w.o.object)
+// openGRPC initializes a pipe for the user to write data to, and a routine to
+// read from that pipe and upload the data to GCS via gRPC.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) openGRPC() error {
+	if err := w.validateWriteAttrs(); err != nil {
+		return err
 	}
-	if !utf8.ValidString(attrs.Name) {
-		return fmt.Errorf("storage: object name %q is not valid UTF-8", attrs.Name)
+
+	pr, pw := io.Pipe()
+	w.pw = pw
+	w.opened = true
+
+	go w.monitorCancel()
+
+	// TODO: Handle this case later.
```

Comment:
I expanded the comment, thanks for the verbiage. 

> writer send the request in a single stream (potentially with multiple messages) using a non-resumable upload (looks like this is also not implemented yet)

This is implemented. See the test `TestIntegration_MultiMessageWriteGRPC`

Reply:
> We don't really have a strategy if the user wants chunks smaller than maxPerMessageWriteSize.

So, I've tested this and it works. The `TestIntegration_MultiChunkWriteGRPC` case sets a `w.ChunkSize` to 1 MB, and i've debugged it (or just run test with `-v` so that it logs progress) to see that it commits 1 MB chunks. The strategy I guess is "write it all in one message and close the stream to commit the chunk", right?

Reply:
Right now our default buffer size is using the `maxPerMessageWriteSize` of 2 MB, but we could use the 16 MB default value that (I think) HTTP has. The code already limits each message written to the `maxPerMessageWriteSize`, so it can split it over multiple messages.

Reply:
Resolved this offline. We need to figure out how to handle the awkward case of `ChunkSize == 0` and not being able to actually do a zero-copy/bufferless upload. This is a TODO.